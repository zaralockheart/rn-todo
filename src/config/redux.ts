// Imports: Dependencies
import {applyMiddleware, combineReducers, createStore} from 'redux'
import thunk from 'redux-thunk'
import AsyncStorage from '@react-native-community/async-storage'
import {createLogger} from 'redux-logger'
import {persistReducer, persistStore} from 'redux-persist'
import {reducer as formReducer} from 'redux-form'

// Imports: Reducers

// Redux: Root Reducer
const rootReducer = combineReducers({
	form: formReducer
});


// Middleware: Redux Thunk (Async/Await)
const middleware: any = [thunk];

// Middleware: Redux Logger (For Development)
if (process.env.NODE_ENV !== 'production') {
	middleware.push(createLogger());
}

// Middleware: Redux Persist Config
const persistConfig = {
	// Root?
	key: 'root',
	// Storage Method (React Native)
	storage: AsyncStorage,
	// Whitelist (Save Specific Reducers)
	whitelist: [
		'authReducer',
	],
	// Blacklist (Don't Save Specific Reducers)
	blacklist: [],
};

// Middleware: Redux Persist Persisted Reducer
const persistedReducer = persistReducer(persistConfig, rootReducer)

// Redux: Store
const store = createStore(
	persistedReducer,
	applyMiddleware(...middleware),
);

// Middleware: Redux Persist Persister
let persistor = persistStore(store);

// Exports
export {
	store,
	persistor,
};
