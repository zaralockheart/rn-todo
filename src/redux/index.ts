export * from './main'
export * from './todo'
