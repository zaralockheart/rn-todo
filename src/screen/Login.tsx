import React from 'react'
import {ErrorOther, Field, InjectedFormProps, reduxForm} from 'redux-form'
import {SafeAreaView, Text, TouchableOpacity} from "react-native"
import TextInput from '../components/form'
import {isNotEmpty} from "../util/string_util"
import {useNavigation} from "react-navigation-hooks"

// import {useNavigation} from 'react-navigation-hooks'

interface LoginProps {
	email: string
	password: string
}

const LoginForm = (props: InjectedFormProps<LoginProps>) => {
	const {handleSubmit} = props

	const {navigate} = useNavigation();

	const validateLogin = ({email, password}: LoginProps) => {
		navigate('Home');
	}

	return (
		<SafeAreaView>
			<Text>Email:</Text>
			<Field name="email" component={TextInput}/>
			<Field name="password" component={TextInput}/>
			<TouchableOpacity onPress={handleSubmit(validateLogin)}>
				<Text>Submit</Text>
			</TouchableOpacity>
		</SafeAreaView>
	)
}

export default reduxForm<LoginProps>({
	form: 'LOGIN_FORM',
	validate(values: LoginProps, props: {} & InjectedFormProps<LoginProps, {}, string>): { [P in keyof LoginProps]?: React.ReactElement | string } & ErrorOther<string> {
		const errors: Partial<LoginProps> = {
			email: undefined,
			password: undefined,
		};
		errors.email = isNotEmpty(values.email)

		errors.password = isNotEmpty(values.password)

		return errors;
	}
})(LoginForm)
