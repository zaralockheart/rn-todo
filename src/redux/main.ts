// Imports: Dependencies
import {applyMiddleware, combineReducers, createStore, Store} from 'redux'
import thunk from 'redux-thunk'
import AsyncStorage from '@react-native-community/async-storage'
import {createLogger} from 'redux-logger'
import {persistReducer, persistStore} from 'redux-persist'
import {FormStateMap, reducer as formReducer} from 'redux-form'
import {composeWithDevTools} from 'redux-devtools-extension'

import {todo, todoState} from './todo'

interface Actions<PROPS> {
    type: string
    payload: PROPS
}

interface ReducersAugment {
    form: FormStateMap
    todo: todoState
}

// Redux: Root Reducer
const rootReducer = combineReducers({
    form: formReducer,
    todo
})

// Middleware: Redux Thunk (Async/Await)
const middleware: [any] = [thunk]

// Middleware: Redux Logger (For Development)
if (process.env.NODE_ENV !== 'production') {
    middleware.push(createLogger())
}

// Middleware: Redux Persist Config
const persistConfig = {
    // Root?
    key: 'root',
    // Storage Method (React Native)
    storage: AsyncStorage,
    // Whitelist (Save Specific Reducers)
    whitelist: ['todo'],
    // Blacklist (Don't Save Specific Reducers)
    blacklist: []
}

// Middleware: Redux Persist Persisted Reducer
const persistedReducer = persistReducer(persistConfig, rootReducer)

// Redux: Store
export type State = ReducersAugment

const store: Store<State> = createStore(
    persistedReducer,
    composeWithDevTools(applyMiddleware(...middleware))
)

// Middleware: Redux Persist Persister
let persistor = persistStore(store)

// Exports
export {middleware, store, persistor, Actions, rootReducer}
