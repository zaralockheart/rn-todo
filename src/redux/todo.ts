import {Dispatch} from 'redux'
import uuidv5 from 'uuid/v5'

import {Actions} from './main'

// Action Names
const ADD_TODO = 'ADD_TODO'
const REMOVE_TODO = 'REMOVE_TODO'
const EDIT_TODO = 'EDIT_TODO'

// Actions
interface AddTodoProps {
  title: string
  subtitle?: string | undefined
  days: {[key: number]: string}
}

const addTodo = (data: AddTodoProps) => (dispatch: Dispatch) => {
  dispatch({
    type: ADD_TODO,
    payload: data
  })
}

interface EditTodoProps extends AddTodoProps {
  id: string
}

const editTodo = (data: EditTodoProps) => (dispatch: Dispatch) => {
  dispatch({
    type: EDIT_TODO,
    payload: data
  })
}

interface RemoveTodoProps {
  id: string
}

const removeTodo = (data: RemoveTodoProps) => (dispatch: Dispatch) => {
  dispatch({
    type: REMOVE_TODO,
    payload: data
  })
}

// All Props
type TodoProps = AddTodoProps & RemoveTodoProps

// Initial State
interface TodoObject {
  id: string
  title: string
  subtitle: string | undefined
  days: {[key: number]: string}
}

const initialState = {
  todo: {} as {[key: string]: TodoObject},
  ids: [] as string[]
}

type todoState = typeof initialState

const todo = (state = initialState, action: Actions<TodoProps>) => {
  switch (action.type) {
    case ADD_TODO: {
      const newId = !!state.ids ? state.ids.slice() : []
      let createdId = uuidv5(Date.now().toLocaleString(), uuidv5.DNS)
      newId.push(createdId)
      const newTodo: {[key: string]: TodoObject} = {
        ...state.todo
      }

      newTodo[createdId] = {
        id: createdId,
        subtitle: action.payload.subtitle,
        title: action.payload.title,
        days: action.payload.days
      }

      const newState: todoState = {
        ids: newId,
        todo: newTodo
      }
      return newState
    }
    case REMOVE_TODO: {
      const newIds = state.ids.slice().filter(i => i !== action.payload.id)

      const newTodo = {
        ...state.todo
      }

      delete newTodo[action.payload.id]
      const newState: todoState = {
        ids: newIds,
        todo: newTodo
      }
      return newState
    }
    case EDIT_TODO: {
      const newTodo = {
        ...state.todo
      }

      newTodo[action.payload.id] = {
        id: action.payload.id,
        title: action.payload.title,
        subtitle: action.payload.subtitle,
        days: action.payload.days
      }

      const newState: todoState = {
        ids: state.ids,
        todo: newTodo
      }

      return newState
    }
    default:
      return state
  }
}

export {TodoProps, TodoObject, todoState, todo, addTodo, removeTodo, editTodo}
