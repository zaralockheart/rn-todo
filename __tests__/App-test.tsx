/**
 * @format
 * @jest-environment jsdom
 */

import React from 'react'
import 'react-native'
import TestRenderer from 'react-test-renderer'
// Note: test renderer must be required after react-native.
import {combineReducers, createStore} from 'redux'
import {Provider} from 'react-redux'
import {Field, reducer as formReducer} from 'redux-form'

import Login from '../src/screen/Login'

const rootReducer = combineReducers({
  form: formReducer
})

const store = createStore(rootReducer)

describe('Login Form', () => {
  it('renders correctly', () => {
    const component = TestRenderer.create(
      <Provider store={store}>
        <Login />
      </Provider>
    )

    expect(component).toMatchSnapshot()

    expect(component.root.findAllByType(Field)).toHaveLength(2)
  })
})
