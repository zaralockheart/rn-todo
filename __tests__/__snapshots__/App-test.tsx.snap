// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Form renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
>
  <Text>
    Email:
  </Text>
  <View>
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      rejectResponderTermination={true}
      style={
        Object {
          "borderColor": "black",
          "borderWidth": 1,
          "height": 37,
          "width": 250,
        }
      }
      testID="email"
      underlineColorAndroid="transparent"
      value=""
    />
    <Text>
      Required
    </Text>
  </View>
  <View>
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      rejectResponderTermination={true}
      style={
        Object {
          "borderColor": "black",
          "borderWidth": 1,
          "height": 37,
          "width": 250,
        }
      }
      testID="password"
      underlineColorAndroid="transparent"
      value=""
    />
    <Text>
      Required
    </Text>
  </View>
  <View
    accessible={true}
    clickable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Text>
      Submit
    </Text>
  </View>
</RCTSafeAreaView>
`;
